### ISSUE TITLE
Protocol matching sometimes depends on order of protocol class method definitions

### ISSUE THREAD

**Issue by timabroad** [microsoft/pyright#9253]
**Describe the bug**
I've got a few classes that are used to define some schemas for serialisation and deserialisation, and have mostly got typing support working, but have come across a curious case where pyright fails to infer a type correctly depending on the order of method definitions in a protocol. MyPy seems to give the expected result in all cases.

**Code or Screenshots**
```python
from __future__ import annotations
from typing import Generic, TypeVar, Protocol, reveal_type

ElemT = TypeVar('ElemT')
OutputT = TypeVar('OutputT')
ElemOutputT = TypeVar('ElemOutputT')

# This function demonstrates/tests the problem.
def test(broken1: MyListBROKEN1[MyInt], broken2: MyListBROKEN2[MyInt], works: MyListWORKS[MyInt]):
    result_broken1 = broken1.deserialise(b'')
    result_broken2 = broken2.deserialise(b'')
    result_works = works.deserialise(b'')

    reveal_type(result_broken1) # Expect list[int].
    reveal_type(result_broken2) # Expect list[int].
    reveal_type(result_works)   # Expect list[int].

# Definition of a class that serialises to a built-in int.
class MyInt:
    @classmethod
    def deserialise(cls, buffer: bytes) -> int:
        ...
    @classmethod
    def serialise(cls, value: int) -> bytes:
        ...

# This protocol is used to capture the input/output types for serialisation.
class Serialisable(Protocol[OutputT]):
    @classmethod
    def deserialise(cls, buffer: bytes) -> OutputT:
        ...
    @classmethod
    def serialise(cls, value: OutputT) -> bytes:
        ...

# This protocol matches a list's element type and then captures what that
# type serialises to.
class SerialisableListWORKS(Protocol[ElemOutputT]):
    @classmethod
    def deserialise(cls, buffer: bytes) -> list[ElemOutputT]:
        ...
    @classmethod
    def serialise(cls, value: list[ElemOutputT]) -> bytes:
        ...
    @staticmethod
    def _element_type() -> Serialisable[ElemOutputT]:
        ...

# This is the same as above, except the _element_type method definition
# comes first.
class SerialisableListBROKEN1(Protocol[ElemOutputT]):
    @staticmethod
    def _element_type() -> Serialisable[ElemOutputT]:
        ...
    @classmethod
    def deserialise(cls, buffer: bytes) -> list[ElemOutputT]:
        ...
    @classmethod
    def serialise(cls, value: list[ElemOutputT]) -> bytes:
        ...

# And this is the same except we're just inheriting the Serialisable
# protocol instead of redefining those methods.
class SerialisableListBROKEN2(Serialisable[list[ElemOutputT]], Protocol[ElemOutputT]):
    @staticmethod
    def _element_type() -> Serialisable[ElemOutputT]:
        ...

class MyListWORKS(Generic[ElemT]):
    @staticmethod
    def _element_type() -> ElemT:
        ...
    @classmethod
    def deserialise(cls: type[SerialisableListWORKS[ElemOutputT]], buffer: bytes) -> list[ElemOutputT]:
        ...
    @classmethod
    def serialise(cls: type[SerialisableListWORKS[ElemOutputT]], value: list[ElemOutputT]) -> bytes:
        ...

class MyListBROKEN1(Generic[ElemT]):
    @staticmethod
    def _element_type() -> ElemT:
        ...
    @classmethod
    def deserialise(cls: type[SerialisableListBROKEN1[ElemOutputT]], buffer: bytes) -> list[ElemOutputT]:
        ...
    @classmethod
    def serialise(cls: type[SerialisableListBROKEN1[ElemOutputT]], value: list[ElemOutputT]) -> bytes:
        ...

class MyListBROKEN2(Generic[ElemT]):
    @staticmethod
    def _element_type() -> ElemT:
        ...
    @classmethod
    def deserialise(cls: type[SerialisableListBROKEN2[ElemOutputT]], buffer: bytes) -> list[ElemOutputT]:
        ...
    @classmethod
    def serialise(cls: type[SerialisableListBROKEN2[ElemOutputT]], value: list[ElemOutputT]) -> bytes:
        ...
```

Pyright output:
```
test.py
  test.py:10:30 - error: Cannot access attribute "deserialise" for class "MyListBROKEN1[MyInt]"
    Could not bind method "deserialise" because "type[MyListBROKEN1[MyInt]]" is not assignable to parameter "cls"
      "MyListBROKEN1[MyInt]" is incompatible with protocol "SerialisableListBROKEN1[ElemOutputT@deserialise]"
      Type "type[MyListBROKEN1[MyInt]]" is not assignable to type "type[SerialisableListBROKEN1[ElemOutputT@deserialise]]"
        "deserialise" is an incompatible type
          Type "(buffer: bytes) -> list[ElemOutputT@deserialise]" is not assignable to type "(buffer: bytes) -> list[ElemOutputT@SerialisableListBROKEN1]"
            Function return type "list[ElemOutputT@deserialise]" is incompatible with type "list[ElemOutputT@SerialisableListBROKEN1]"
        Could not bind method "serialise" because "type[MyListBROKEN1[MyInt]]" is not assignable to parameter "cls"
          "MyListBROKEN1[MyInt]" is incompatible with protocol "SerialisableListBROKEN1[ElemOutputT@serialise]" (reportAttributeAccessIssue)
  test.py:11:30 - error: Cannot access attribute "deserialise" for class "MyListBROKEN2[MyInt]"
    Could not bind method "deserialise" because "type[MyListBROKEN2[MyInt]]" is not assignable to parameter "cls"
      "MyListBROKEN2[MyInt]" is incompatible with protocol "SerialisableListBROKEN2[ElemOutputT@deserialise]"
      Type "type[MyListBROKEN2[MyInt]]" is not assignable to type "type[SerialisableListBROKEN2[ElemOutputT@deserialise]]"
        "deserialise" is an incompatible type
          Type "(buffer: bytes) -> list[ElemOutputT@deserialise]" is not assignable to type "(buffer: bytes) -> list[ElemOutputT@SerialisableListBROKEN2]"
            Function return type "list[ElemOutputT@deserialise]" is incompatible with type "list[ElemOutputT@SerialisableListBROKEN2]"
        Could not bind method "serialise" because "type[MyListBROKEN2[MyInt]]" is not assignable to parameter "cls"
          "MyListBROKEN2[MyInt]" is incompatible with protocol "SerialisableListBROKEN2[ElemOutputT@serialise]" (reportAttributeAccessIssue)
  test.py:14:17 - information: Type of "result_broken1" is "Unknown"
  test.py:15:17 - information: Type of "result_broken2" is "Unknown"
  test.py:16:17 - information: Type of "result_works" is "list[int]"
2 errors, 0 warnings, 3 informations
```

Mypy output:
```
test.py:14: note: Revealed type is "builtins.list[builtins.int]"
test.py:15: note: Revealed type is "builtins.list[builtins.int]"
test.py:16: note: Revealed type is "builtins.list[builtins.int]"
```

**VS Code extension or command-line**
Observed in VSCode (pylance v2024.10.1) and with pyright 1.1.385 on the command-line.

**Comment by timabroad** [comment]
Also just noticed that if you change the `Serialisable` protocol class to be covariant in `OutputT` by removing the `serialise()` definition, that makes pyright work correctly regardless of ordering. Might help track down the issue? Or just be a red herring 😆!

**Comment by erictraut** [comment]
I agree this is a bug. Thanks for reporting it.

This will be addressed in the next release.

While investigating the problem, I noticed something suspect in your code. The static method `_element_type` is returning a value defined using a class-scoped type variable. I don't see how this can work in practice because type arguments are erased at runtime. I'm curious how you intended for this to work. In particular, how would one implement this method?

**Comment by timabroad** [comment]
Ah, that's just a hack to make the typing work - it isn't actually implemented or intended to be used (hence the underscore-prefixed name to flag it as private). The problem is that we need a way for `ElemT` in `MyList` to be matched with `Serialisable[ElemOutputT]` in `SerialisableList`, so that it can "extract" the type `OutputT` from `ElemT`. If you get rid of that definition, then the only functions on the Protocol that line up just end up with circular matching of `list[ElemOutputT]` to `list[ElemOutputT]`, so the types all reveal as `list[ElemOutputT@serialise]`. Maybe there's a better way to make this work?

Having said that... I do actually have some other code that is doing something similar, which runs into some issues with pyright complaining that I can't do what I'm doing 😆 - but I think that they're outside of the typing spec so I just ignore it any carry on. There are a couple of ways you can make this work in practice though, which basically boil down to having your base type as not intended to be instantiated directly, but just being a placeholder superclass to hold the Generic.

For a simple example:
```python
class MyListBase(Generic[T]):
    @classmethod
    @abstractmethod
    def get_type(cls) -> type[T]: ...

class MyIntList(MyListBase[int]):
    @classmethod
    def get_type(cls) -> type[int]:
        return int

class MyStrList(MyListBase[str]):
    @classmethod
    def get_type(cls) -> type[str]:
        return str
```

This example feels like it potentially should be supported - in fact pyright doesn't complain about anything in this, but if you change from a `@classmethod` to a `_type: ClassVar[type[T]]` you get an error "ClassVar type cannot include type variables" even if you leave it unassigned, but it will correctly typecheck `MyIntList` and `MyStrList` when you try to assign to it (eg if you set `_type = float` or `_type = 17` pyright knows that this is wrong). But the original PEP specifically disallows these classvars, so I just break the rules and carry on 😆.

I've also got some code that overrides `__class_getitem__` specifically to create non-erased runtime types for some things. I don't really know if this is supposed to be supported or not, but I just crack on with it since it mostly works (bar the `ClassVar` issue, and currently this bug) 😃.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.386

**Comment by timabroad** [comment]
Awesome, thanks for the rapid response!
